src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -78,9 +78,21 @@ internal static bool IsEnabled()
             // If we are on at all, we propagate any activity information
             // unless tracing system or user injected Request-Id for backward compatibility reasons.
             Activity currentActivity = Activity.Current;
-            if (currentActivity != null && !request.Headers.Contains(DiagnosticsHandlerLoggingStrings.RequestIdHeaderName))
+            if (currentActivity != null)
             {
-                request.Headers.Add(DiagnosticsHandlerLoggingStrings.RequestIdHeaderName, currentActivity.Id);
+                if (currentActivity.IdFormat == ActivityIdFormat.W3C && !request.Headers.Contains(DiagnosticsHandlerLoggingStrings.TraceParentHeaderName))
+                {
+                    request.Headers.Add(DiagnosticsHandlerLoggingStrings.TraceParentHeaderName, currentActivity.Id);
+                    if (currentActivity.TraceStateString != null)
+                    {
+                        request.Headers.Add(DiagnosticsHandlerLoggingStrings.TraceStateHeaderName, currentActivity.TraceStateString);
+                    }
+                }
+                else if(!request.Headers.Contains(DiagnosticsHandlerLoggingStrings.RequestIdHeaderName))
```

Comment:
Do you want the 'if' condition to have two terms in it?   
As it is if the ID format is W3C AND you already have a TraceParent header, you ALSO try to add a RequestId header.   

This is not what I would have expected.   Feels like what you want is 
```
                if (currentActivity.IdFormat == ActivityIdFormat.W3C)
                {
                    if (request.Headers.Contains(DiagnosticsHandlerLoggingStrings.TraceParentHeaderName))
                    {
                        request.Headers.Add(DiagnosticsHandlerLoggingStrings.TraceParentHeaderName, currentActivity.Id);
                        if (currentActivity.TraceStateString != null)
                        {
                            request.Headers.Add(DiagnosticsHandlerLoggingStrings.TraceStateHeaderName, currentActivity.TraceStateString);
                        }
                    }
                }
                else // Not W3C Format
                { 
                    if(!request.Headers.Contains(DiagnosticsHandlerLoggingStrings.RequestIdHeaderName))
                    {
                        request.Headers.Add(DiagnosticsHandlerLoggingStrings.RequestIdHeaderName, currentActivity.Id);
                    }
                }
```

Reply:
But otherwise it looks OK.

Reply:
thanks, fixed